llvm/lib/Target/SPIRV/SPIRVTypeRegistry.cpp

```diff
@@ -387,6 +387,93 @@ SPIRVTypeRegistry::buildConstantSampler(Register resReg, unsigned int addrMode,
   return res->getOperand(0).getReg();
 }
 
+bool
```

Comment:
Why moving this? I personally don't like the idea of TR doing anything then handling types, I was planning to strip it as much as possible, but maybe there's some rationale behind this?

Reply:
It's gathering of OpDecorate and OpVariable builders in one place to avoid code duplication.

For now I see the another appropriate place for OpDecorate generation which is SPIRVStrings. It's logically close to buildOpName and also uses addStringImm so I'll move it there. Perhaps we should create SPIRVUtils.* for this kind  of generic functions, if we have more of them.

What about OpVariable building, I would propose to gather all DT.add/DT.find users in one place and cover them by interfaces similar to those we have for types e.g. getOrCreateType/Constant/GlobalVariable and maybe Function. Now the builder for Constants is already in TR, and after the patch buildGlobalVariables will be there.

As for having only Type-related things in TR, we could split it into TypeRegister and ObjectRegister, or even ConstRegister and GVRegister instead of the Object's one. However as for me, it's convenient to use one pointer like TR to generate all types of types/consts/gvs in different places of the backend,  instead of making 2 or 3 such pointers.  So probably we'd better use TypeRegister as GlobalEntityRegister, and maybe rename it to GlobalRegister or something similar.

Some code of SPIRVTypeRegistry actually can be moved to other places, e.g. getTypeFoldingSupportingOpcs, isTypeFoldingSupported, perhaps type mangling, propagateSPIRVType, insertAssignInstr, generateAssignInstrs. The last three probably should go to IRTranslator.

Reply:
Fair point. I still think we need to elaborate on the overall design of this stuff before doing any refactoring. I suppose it's not this movement which improves passrate in this PR, is it?

Reply:
> I suppose it's not this movement which improves passrate in this PR, is it?

Actually proper tracking of OpVariables in DT during builtins call lowering affects the pass rate and makes improvements on some other tests. I did not want to duplicate selectGlobalVariable code into SPIRVOpenCLBIFs.cpp, so I put it to TR since it's easy accessible from SPIRVOpenCLBIFs.cpp and SPIRVInstructionSelector.cpp. I suppose it's not a problem for now to have buildGlobalVariable in TR, isn't it?

> For example I have plans to introduce some general interface...

Nice plan! Although the sorting could be done even without keeping in the same storage, one storage looks universally.
By the way continuing the plan for "GlobalRegister" that I described before, I would incapsulate this unified DT storage into the "GlobalRegister".

Reply:
I have updated the patch moving buildOpDecorate into SPIRVStrings* and correcting casts.